src/shared/sam/cli/samCliInvoker.ts

```diff
@@ -8,42 +8,168 @@
+    SamCliVersionValidation,
+    SamCliVersionValidator,
+    SamCliVersionValidatorResult
+} from './samCliVersionValidator'
 
-export interface SamCliProcessInvoker {
-    invoke(options: SpawnOptions, ...args: string[]): Promise<ChildProcessResult>
-    invoke(...args: string[]): Promise<ChildProcessResult>
+interface SamCliProcessInvokerContext {
+    cliConfig: SamCliConfiguration
+    cliInfo: SamCliProcessInfo
+    cliUtils: SamCliUtils
+    validator: SamCliVersionValidator
 }
 
-export interface SamCliTaskInvoker {
-    invoke(task: vscode.Task): Promise<vscode.TaskExecution>
+export function makeSamCliProcessInvokerContext(
+    params?: Partial<SamCliProcessInvokerContext>
+): SamCliProcessInvokerContext {
+    if (!params) {
+        params = {}
+    }
+
+    return {
+        cliConfig: params.cliConfig || new DefaultSamCliConfiguration(
+            new DefaultSettingsConfiguration(extensionSettingsPrefix),
+            new DefaultSamCliLocationProvider()
+        ),
+        cliInfo: params.cliInfo ||  { info: undefined, lastModified: undefined },
+        cliUtils: params.cliUtils || new DefaultSamCliUtils(),
+        validator: params.validator || new DefaultSamCliVersionValidator(),
+    }
 }
 
 export class DefaultSamCliProcessInvoker implements SamCliProcessInvoker {
-    public constructor(private readonly config: SamCliConfiguration = new DefaultSamCliConfiguration(
-        new DefaultSettingsConfiguration(extensionSettingsPrefix),
-        new DefaultSamCliLocationProvider()
-    )) {
-    }
+
+    public constructor(private readonly _context: SamCliProcessInvokerContext) {}
 
     public invoke(options: SpawnOptions, ...args: string[]): Promise<ChildProcessResult>
     public invoke(...args: string[]): Promise<ChildProcessResult>
     public async invoke(first: SpawnOptions | string, ...rest: string[]): Promise<ChildProcessResult> {
-        const args = typeof first === 'string' ? [ first, ...rest ] : rest
-        const options: SpawnOptions | undefined = typeof first === 'string' ? undefined : first
+        const logger: Logger = getLogger()
 
-        const samCliLocation = this.config.getSamCliLocation()
+        const samCliLocation = this._context.cliConfig.getSamCliLocation()
         if (!samCliLocation) {
-            throw new Error('SAM CLI location not configured')
+            const err = new Error('SAM CLI location not configured')
+            logger.error(err)
+            throw err
+        }
+        const validationResult = await this.getCliValidation(samCliLocation)
+
+        if (validationResult.validation === SamCliVersionValidation.Valid) {
+            const args = typeof first === 'string' ? [ first, ...rest ] : rest
+            const options: SpawnOptions | undefined = typeof first === 'string' ? undefined : first
+
+            return await this.runCliCommand(samCliLocation, options, ...args)
         }
 
+        const errorResult: ChildProcessResult = {
```

Comment:
`ChildProcessResult` is intended to represent the raw results of a child process. So `exitCode` should be the child process's exit code, and `error` be any error thrown by running the child process.

Instead, suggest to use the default `Error` type, or your own extension of this type (possibly with a `validationResult` property).